### ISSUE TITLE
We should consider using toggles instead of checkboxes to represent direct manipulation of boolean settings

### ISSUE THREAD

**Issue by lampholder** [vector-im/riot-web#5392]
https://uxplanet.org/checkbox-and-toggle-in-forms-f0de6086ac41

As we move towards direct manipulation settings modification paradigm across the board, we should use toggles instead of checkboxes.

We should also use positive wording (e.g. `[ ] Send typing notifications`, rather than `[X] Don't send typing notifications` to aid easy comprehension).

Caveat - checkboxes are sometimes better when we have to take further steps to action the change, so this change shouldn't be applied without thought to the specific use case.

**Comment by uhoreg** [comment]
Hmm.  Am I the only one who doesn't like toggles?  They're usually at least twice as wide as checkboxes.  Unless you have labels (which as the article states, should be external, and then take up more space), it's unclear which side is "on" and which side is "off".  The only real indication of the state is colour, which can be problematic for people with colour blindness.

Checkboxes are compact, clear, rely on shape rather than colour, and are native controls in browsers and hence is guaranteed to have accessibility features baked in.

Also, I'm not sure that "direct action" vs "requires further action" is a good way to distinguish between when to use a checkbox vs a toggle, since no other control has separate versions for those modes of operation.

But +1 to using positive wording.

**Comment by pafcu** [comment]
It's possible to style a Checkbox like a toggle, so losing accessibility should not be a problem. As to color blindness, it's extremely rare that someone would not actually be able to distinguish between e.g. grey and green. Colorblind people don't literally see in black and white.

The width of a toggle is not necessarily a problem, it actually makes it easier to hit it. It's also not like there is a lack of vertical space in settings, almost the opposite.

The fact that other controls don't have versions that clearly state how they behave is IMHO a problem of other controls, not an argument for using checkboxes.

A nice feature of toggles is that they are similar to sliders, so you can also easily move to a nice "Off - On - Noisy" slider for notifications.

**Comment by uhoreg** [comment]
> Colorblind people don't literally see in black and white.

It depends on the type of colour blindness.  Red-green colour blindness is the most common, but there are some people who do effectively see in black and white.

> The width of a toggle is not necessarily a problem, it actually makes it easier to hit it.

If the label is used as a target too (as it should be), then the additional size doesn't make it significantly easier to hit.

> It's also not like there is a lack of horizontal space in settings, almost the opposite.

That's a fair enough point.  Confusingly to me, toggles are popular on mobile devices, where horizontal space is less plentiful.

> The fact that other controls don't have versions that clearly state how they behave is IMHO a problem of other controls, not an argument for using checkboxes.

It's more that, since other controls don't have different versions, you need to rely on something else to indicate the mode of the control.  And I think that a UI that uses checkboxes in some places and toggles in other places would look confused.

**Comment by pafcu** [comment]
>  Red-green colour blindness is the most common, but there are some people who do effectively see in black and white.

Fair enough, but it is something like 1 in 33000 people. Even then, these people are able to see different shades of gray and tell if something is light or dark. Now, that of course does not tell whether it is on or not, but neither does something being green or white, or something having a checkmark or not. You learn UI conventions. And in any case, this only applies if there are no labels.

> If the label is used as a target too (as it should be), then the additional size doesn't make it significantly easier to hit.

Look at normal user interactions and notice how people try to hit the checkboxes anyway. Even I tend to click the checkboxes even though _I know_ clicking on the label works, because I made it that way myself!

>  And I think that a UI that uses checkboxes in some places and toggles in other places would look confused.

I think that is more a symptom of having a UI that needs settings to be explicitly saved in some places, and not in others.

**Comment by uhoreg** [comment]
> Now, that of course does not tell whether it is on or not, but neither does something being green or white, or something having a checkmark or not. You learn UI conventions.

Checkboxes are a UI convention that have been used, and were quite common, since before computers were in use, so it's very well-known.  The UI conventions for toggles are much less well-established.  (If toggles were vertical instead of horizontal, that might be a different story, since it would be a closer approximation to light switches, but horizontal switches are less common.)

> And in any case, this only applies if there are no labels.

Adding labels makes it so that users have even more to read.

Basically, toggles [make me think](https://en.wikipedia.org/wiki/Don't_Make_Me_Think) more than checkboxes do, instead of letting me focus on the task I want to do.

**Comment by pafcu** [comment]
> Checkboxes are a UI convention that have been used, and were quite common, since before computers were in use, so it's very well-known.

I believe that is why people avoid them nowadays. They are associated with the old, explicit-save UIs, as compared with the more modern save-less approach.

In any case, my opinion is not important. I'll defer to research and UX experts.

**Comment by turt2live** [comment]
With respect the uncertainty of "off vs on", the convention is right side = on, left side = off. Anything that doesn't follow this convention is asking for trouble.

**Comment by uhoreg** [comment]
> With respect the uncertainty of "off vs on", the convention is right side = on, left side = off.

Yup, that's the convention, but even though it's the convention, I still have to think about it every time, because it isn't intuitively obvious which should be which, and it's easier to mix up.  Whereas with a checkbox, it's clear what's what and harder to mix up.

**Comment by lampholder** [comment]
Whatever we end up doing, our driving objective is to minimise user cognitive load (as per https://en.wikipedia.org/wiki/Don't_Make_Me_Think). It's with this in mind that we're trying to steer (albeit slowly) towards a single coherent settings paradigm.

> And I think that a UI that uses checkboxes in some places and toggles in other places would look confused.

100% agreed - but our settings UI today _is_ confused; hiding the confusion with cosmetic similarity is arguably not a good thing :grin:  

IMHO getting settings right is v. important, so I think the thing to do is mock up some options and get them in front of a range of people for feedback (when we have the bandwidth to do such a thing, of course).

Title modified to be less presumptive of our eventual conclusions ;)

**Comment by pvagner** [comment]
> It's possible to style a Checkbox like a toggle using CSS, so losing accessibility should not be a problem.

Modern browsers with up to date assistive tools have great support for both traditional checkboxes and toggle buttons. I would say that we *should* try to apply universal design where possible and try to make as less differences to the ux for different people as possible. For example for me toggle button is a toggle button and a checkbox is a checkbox. They can be both toggled by pressing the spacebar key however I would still like to have them clearly differentiated because mixed implementations lead to more support questions.
When I have to describe part of the UI to someone else I like to use the semantics where available because I can't rely on visual appearance as a blind individual. Thus I tend to say something like:
- Open Riot
- Use the arrow keys or tab /shift+tab to navigate over the buttons found inside the complementary content
- Activate the Settings button there
- Navigate by headings once the settings dialog loads until you find URL previews heading
- Now press tab key to find a checkbox.

Boom, I assume it's a checkbox because it's a checkbox internally however you have styled it as a toggle button thus if I am saying locate a checkbox, very few people will be able to actually find it.
Of course this is not black and white and in order to provide great accessibility sometimes relying on traditional controls is a clear win. However for simple two state controls, buttons, radio buttons, switches, tab controls, dropdown buttons and similar UI elements there is no excuse to try to make experience equal for both average users as well as those relying on assistive tools.

**Comment by pafcu** [comment]
@pvagner: Interesting. However, how do you know what a toggle is in that case, as there is no semantic element for it? Often I see toggles implemented as e.g. a div, or even a normal button, with some Javascript attached to it, and that seems even less accessible than styling a checkbox.

**Comment by pvagner** [comment]
Reading this discussion now ilustrates how we are looking at this from an opposite sides. Originally this discussion started by saying toggle buttons are prettier and very fashioned these days allowing even more visual appearance tweaks than checkboxes although not all the people may like it.
Yesterday when writing my reply I was not looking at how you might be able to implement this I have simply remembered I have seen lots of toggle buttons on Facebook, Gmail, possibly Twitter, Google drive, Microsoft Office online and other rich web apps that are known to be accessible. And in deed I have noticed the difference so much that it prompted me to write that comment yesterday. Without further checking I assumed there is some toggle button HTML5 element because I can see increasing number of toggle buttons on the web.
Now I've checked and there is no native toggle button HTML input element yet at least I am unable to find it. However [there are some easy to read docs](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_button_role) on how to go about adding some accessibility related juice to the custom toggle buttons.
And now I think we can try to make some good use of it. In fact majority of simple buttons and image buttons are already implemented like this in Riot thanks to @JaniM and @ara4n. So by commenting here I just would like to make sure if you are considering changing checkboxes to custom toggle buttons, make sure these are exposed as toggle buttons to assistive tools and they are useable with the keyboard alone.
You can see these are really minor tweaks but as the visual appearance matters this also positivelly affects look and feel when using the app with screen readers and similar.
Another very simple tweak would be adding aria-pressed property to the right pannel Members, Files and Notification buttons according to the visibility of their corresponding pannels.
Another more but still very simple markup would be exposing This room, All rooms in the Search structure as radio buttons by using [role=radio](https://www.w3.org/TR/wai-aria-1.1/#radio) and if there is an encapsulating parent element such as div it should be exposed as a [radiogroup](https://www.w3.org/TR/wai-aria-1.1/#radiogroup).
Then we need repeat what @JaniM has done some 9 months ago to bring the accessibility support to the new UI parts such as message input editor, call buttons and similar.
There is also easy markup for simple [dialogs](https://www.w3.org/TR/wai-aria-1.1/#alertdialog) or [status messages](https://www.w3.org/TR/wai-aria-1.1/#alert) requiring typically no changes or in some cases very minimal changes to the existing code. By the way if an error popup comes up on Riot I am only presented with OK button and I need to read its surounding text in order to discover what has gone wrong. Try to imagine how much polished the experience can get with such minor adjustments if I was automagically alerted when these things are happening.
If you like we can talk more about this, I have some rough ideas on what's good when implementing accessibility, I do also know what might be usefull however I am still not clever developer and since I am blind I tend to fear I might inadvertently hinder look and feel when I am not working in a team where someone might quickly look at my stuff.

**Comment by pafcu** [comment]
@pvagner My idea of simply styling check-boxes was exactly so that keyboard navigation and all the standard accessibility stuff would work _without_ any extra effort, just like they work for check boxes (because they actually are check-boxes): a screen-reader would understand that it is something that toggles a value.

The difference between check-boxes and toggles is not about fashion, but about how they usually work: check-boxes usually need to be explicitly saved, while toggles are saved automatically. There is no real functional difference: both toggle a value. It's just that the visual appearance tell people whether they will be saved immediately or not. This is simply a convention, but people have learned to expect it. Naturally it would be great to be able to expose this difference to screen-readers as well, so adding an aria attribute would be great.

Regarding accessibility in general it might make sense to open a (or several) issues about how it can be specifically improved. For example, you could open an issue and list all the places you think would gain from additional aria attributes.

I personally feel that accessibility is very important, and that software should be usable by everyone, but for me with normal vision and hearing it difficult to find the problematic spots. Perhaps it would be a good idea to create a #riot-accessibilty:matrix.org, just like there is a room for translations?

**Comment by pvagner** [comment]
I consider #302 as a general a11y related issue

**Comment by turt2live** [comment]
During my implementation I have no plans to break accessibility of checkboxes.

**Comment by uhoreg** [comment]
> It's just that the visual appearance tell people whether they will be saved immediately or not. This is simply a convention, but people have learned to expect it.

I'm not sure that's true.  Before reading the article posted above, I for one did not know about any such convention like that.  And the other day I saw a UI where a toggle was used for something that was clearly not going to be saved immediately.  And GitLab uses toggles in one spot with a "Save changes" button, which makes one believe that those toggles do not save immediately.  So it seems like that convention is not as widespread as one might hope.

But it seems like the answer to my original question in this thread is: "Yes, I am the only one who doesn't like toggles."  So <shrug>.

**Comment by uhoreg** [comment]
FWIW, Google's [Material Design guidelines](https://material.io/guidelines/components/selection-controls.html) say that checkboxes should be used for "selection of multiple options from a set", while toggles (which they call "switches") should be used for "a selection to be turned on or off" without making any distinction between whether the setting is applied immediately or not.  The distinction between "selection of multiple options from a set" versus "turning something on/off" makes more sense to me, and I'd prefer using that distinction instead of "saves immediately" versus "saves later".  (Though of course I still dislike toggles. 😛 )

**Comment by turt2live** [comment]
Their guidelines would be more for tables (as they show). A list of distinct settings is switchable under their guidelines.

With granular settings everything is locally echoed, so the boolean settings do get saved right away, alongside a request to persist the setting. If the request ultimately fails then it reverts the change. It shouldn't fail though, given the whole system has checks in place to have good confidence that it's doing the right thing.

The good news is that toggles vs checkboxes is nothing harder than a find&replace battle. If toggles were to go horribly wrong for most users, then it's easy to go back to checkboxes.

**Comment by ara4n** [comment]
this ends up being very subjective, but we're going to give toggles a go (sorry uhoreg).

**Comment by GReagle** [comment]
I hate toggle switches. I don't see any advantage to using them over checkboxes.

Toggle switches definitely take more cognitive load for me than checkboxes.  

I have red-green color blindness, which is a (relatively) common form of it.  I find the color distinction used by toggle switches hard to work with.  It is incorrect to conclude that this is an accessibility issue that applies only to people who see only in black and white.  According to Wikipedia:  "Red–green color blindness affects up to 8% of males and 0.5% of females of Northern European descent."  When I look at a checkbox, it is blindingly obvious to me whether it is checked or not; I don't have to rely on the developers to choose colors that stand out to me as a red-green color-blind person.

"check-boxes usually need to be explicitly saved, while toggles are saved automatically":  This is an interesting convention, but I am not familiar with it.  I look for a "Save Settings" or "OK/Cancel" buttons to figure out whether a setting change will be instant or not, if I care, which I usually don't.  Even if this is a real and common convention, and has the advantage of signalling whether a change in setting will be instant or not, that benefit does not outweigh the other horrible disadvantages of toggle switches, which have been explained well by uhoreg .

The fact that toggle switches are more "modern and pretty" and that checkboxes are considered "old-fashioned" is just BS to me.
